### ISSUE TITLE
Prefer explicit over implicit version in MediaTypeApiVersionReader

### ISSUE THREAD

**Issue by DerNamenlose** [microsoft/aspnet-api-versioning#746]
When having a `POST` request with `Content-Type` and `Accept` headers, the `MediaTypeApiVersionReader` will always prefer the `Content-Type` header, _even if_ it does not contain any version parameter at all. I'd like to propose to always prefer explicit versions over implicit ones, i.e. if the `Accept` header contains a version parameter and the `Content-Type` does not, then the `Accept` header should be preferred.

We have this use case in our product, where the `Content-Type` of the `POST` request body will not change between different versions (it is a rather stable filter object), but the `Accept` header will, because the response objects differ. The current behavior requires us to introduce a new `Content-Type` for the `POST`, even though its actual contents did not change at all.

I would have a pull request with the changed behavior and appropriate tests ready, if there is interest in that.

**Comment by commonsensesoftware** [comment]
There's been a few debates about this before. I've contemplated changing it a few times, but my better judgement keeps coming back. The version _should_ not be thought of as a separate thing; it _is_ part of the media type. Honestly, it's really part of the content negotiation regardless of whether you version by media type. You have elected to version by media type. Congrats on going the RESTful way! 😉 

These headers have very specific meanings:

- `Content-Type` - What I, the client, am giving you server
- `Accept` - What I, the client, want the server to give me back

While it might seem pedantic, the differences are important. These two headers do not have to match - neither in API version nor media type. This is a perfectly valid request:

```http
POST /order HTTP/2
Host: localhost
Accept: text/csv; q=0.6, application/xml; q=0.4, text/plain
Content-Type: application/json
Content-Length: 42

{"name": "John Doe", "products": [42]}
```

Where I'm going with this is that it is more than reasonable to have `Content-Type: application/json; v=1.0` and `Accept: application/json; v=2.0`. This behavior is not supported out-of-the-box, but it's possible to make it work. You would likely need a custom `InputFormatter` and `OutputFormatter` for each API version (or something like that).

Ultimately, the main challenge here is that there is not a universal way to accommodate all APIs that version this way. Changing formatters requires quite a bit of remapping. This is the same reason that the `application/vnd.json+v1` style isn't intrinsically supported. The format to parse it is not standard and you'd have to remap all the formatters.

Making things _just work_ out-of-the-box is a goal, but one I know will never reach 100% for all APIs or styles. This is exactly why there are many hooks to customize and extend things. In your particular case, I would recommend that you either extend the built-in `MediaTypeApiVersionReader` or roll your own custom `IApiVersionReader` that uses the behavior you want. The effort is pretty minimal and you have the existing source to use as a primer. Then you only need wire it up a la:

```c#
options.ApiVersionReader = new PreferAcceptOverContentTypeApiVersionReader();
```

Or something like that.

Community support and pull requests are great, but I have to balance between what makes sense for all and enabling customization for specific implementations.

**Comment by DerNamenlose** [comment]
I see where you're coming from and I actually agree. There has to be some kind of default behavior and preferring `Content-Type` over `Accept` is as good as any. What I'd like to argue, that an *explicit* version should take precedence over an *implicit* one, i.e. if (and only if) the `Accept` header DOES specify a version, while the `Content-Type` DOES NOT, then the `Accept` header should take precedence (which is the behavior, that my pull request implements). Your `Content-Type: ...; v=1.0`, `Accept: ..., v=2.0` example would still work the same way, as the `Content-Type` specifies an explicit version.

In conclusion: it's not about `Content-Type` vs. `Accept`, but about *explicit* vs. *implicit*.

**Comment by commonsensesoftware** [comment]
You are correct and I agree that explicit should always take precedence over implicit. In fact, the design tries really hard to make sure that behavior cannot be broken. If a client asks explicitly asks for version 1, then the server should not arbitrary decide version 42 under any circumstance. To do so breaks the whole premise of versioning.

The behavior isn't so much about `Accept` versus `Content-Type`. I should have elaborated further. The behavior is due to the client sending content in a `POST` and `Content-Type`. The `Content-Type` is the one and only header that can be sent to the server that defines what the body means in accordance with [RFC 7231 §3.1.1.5](https://datatracker.ietf.org/doc/html/rfc7231#section-3.1.1.5).

With these rules, the logic is really:

- Prefer content over no content
- If content is present, then the `Content-Type` header _should_ be preferred (it's explicit)
- If no content is present, fallback to `Accept` (it's explicit) using the requested media type with the highest quality (e.g. `q`)

The one place where these rules did not make sense was for `DELETE`. Neither `Accept` nor `Content-Type` applies or makes sense when used in the context of `DELETE`. Nevertheless, for a long time, one of these two headers (usually `Accept`) needed to be specified. It is now possible to define a _version-neutral_ API down to the action. This is the most appropriate behavior as `DELETE` _shouldn't_ have any input or output and _should_ have no visible changes externally. For example, `DELETE /order/123` is the same for all versions. Defining  `DELETE` as _version-neutral_ means that it will accept (and ignore) any API version (for parity with other APIs), including no API version at all.

This leads me to another edge case. It is possible that someone will map their `POST` action as API _version-neutral_ and decide the response by inspecting the `Accept` header. It's somewhat of a matter of preference for the implementation, but this mapping _could_ even work for your scenario.

Your request and scenario is not unreasonable, I just don't know that it should be the de facto implementation for all.

This all you'd need to do to achieve the desired behavior based on your original request:

```c#
public class ContentTypeOrFallbackToAcceptApiVersionReader : MediaTypeApiVersionReader
{
    protected override string? ReadContentTypeHeader( MediaTypeHeaderValue contentType )
    {
        var result = base.ReadContentTypeHeader( contentType );

        if ( string.IsNullOrEmpty( result ) )
        {
            var accept = request.GetTypedHeaders().Accept;
            return accept is null ? default : ReadAcceptHeader( accept );
        }

        return result;
    }
}

// ...

services.AddApiVersioning( options => options.ApiVersionReader = new ContentTypeOrFallbackToAcceptApiVersionReader() );
```

Ultimately, `Accept` defines the **output** of a request and `Content-Type` defines the **input** of a request. You cannot get to the output without going through the input. In the spirit of explicit vs implicit, I'm not sure that the output format (or a part of it such as the API version) should _override_ or otherwise be implicitly inferred as a fallback to incomplete input.

**Comment by commonsensesoftware** [comment]
Small update on this. Things are changing in `6.0`; specifically, that `IApiVersionReader.Reader` will now return `IReadOnlyList<string>` instead of `string?`. This makes the argument and debate over this topic more or less mute. You can see the new implementation [here](../blob/main/src/AspNetCore/WebApi/src/Asp.Versioning.Http/MediaTypeApiVersionReader.cs). As you can see, there is a clear and straight forward way to grab version from both headers (even if _I believe_ it should be more prescriptive). Note that the request will still fail if both are specified and they are not **exactly** the same.

The only feasible way to deal with different API versions by media type is to have custom **InputFormatter** and **OutputFormatter** per API version. Currently, there isn't a generic way to achieve. _Maybe_ in the future.

**Comment by commonsensesoftware** [comment]
The new behavior and functionality is now available in `6.0` in either:

- [Asp.Versioning.WebApi](https://www.nuget.org/packages/Asp.Versioning.WebApi) (ASP.NET Web API)
- [Asp.Versioning.Http](https://www.nuget.org/packages/Asp.Versioning.Http) (ASP.NET Core)

It wasn't clear which platform you were targeting. The referenced source logic can also easily be backported if you need stay on an earlier version. Thanks for the discussion. I think it advanced what the community is interested in seeing happen.
